cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -958,6 +987,71 @@ func getNotRegisteredNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances ma
 	return notRegistered
 }
 
+// Calculates which of the registered nodes in Kubernetes that do not exist in cloud provider.
+func (csr *ClusterStateRegistry) getCloudProviderDeletedNodes(allNodes []*apiv1.Node, cloudProviderNodeInstances map[string][]cloudprovider.Instance) []*apiv1.Node {
+	nodesRemoved := make([]*apiv1.Node, 0)
+	currentCloudInstances := make(map[string]string, 0)
+	registeredNodes := make(map[string]*apiv1.Node, 0)
+	if len(allNodes) > 0 {
+		_, err := csr.cloudProvider.NodeExists(allNodes[0])
+		// Check if the cloud provider implements nodeExists method
+		nodeExistsNotImplemented := errors.Is(err, cloudprovider.ErrNotImplemented)
+		if nodeExistsNotImplemented {
+			// Fall-back to taint-based node deletion
+			for _, node := range allNodes {
+				if deletetaint.HasToBeDeletedTaint(node) {
+					nodesRemoved = append(nodesRemoved, node)
+				}
+			}
+		} else {
+			for nodeGroupName, instances := range cloudProviderNodeInstances {
+				for _, instance := range instances {
+					currentCloudInstances[instance.Id] = nodeGroupName
+				}
+			}
+			for _, node := range allNodes {
+				registeredNodes[node.Name] = node
+			}
+
+			// Fill previously deleted nodes, if they are still registered in Kubernetes
```

Comment:
One concern I have with the new interface using it to discover deleted nodes.  At the moment, I don't have a clear understanding of the level of effort involved with verifying the instance's existence with each cloud provider.  Since the ClusterStateRegistry stores the previousCloudProviderNodeInstances and cloudProviderNodeInstances, I would like to leverage them to identify nodes that may be removed/terminated on the cloud provider.  Afterwards, the new interface method can be used to verify with the cloud provider that the node exists or not.  This will prevent marking a node as deleted if it was not-autoscaled or removed from a node group.

Unfortunately, I'm not sure how to support a case when a cloud provider might only implement the new interface for specific nodes.  That would require that the check is performed on each node to ensure taint-based deletion is checked for each node.

Are there concerns about looping over `allNodes` and checking each with the new interface method?

Reply:
Unfortunately, I will need to remove this logic since it can't support both the taint-based deletion and new method simultaneously.  The trouble is that the new technique will discover potential nodes that no longer exist between the two caches (`cloudProviderInstances` recently retrieved and the cached data in the clusterStateRegistry).  This would make it tough to also identify the nodes that have the taint, because those instances would not be otherwise known.

As you mentioned, it will be up to each cloud provider to implement an efficient implementation of this `HasInstance` method.  I will modify the behavior here to iterate over the list of `allNodes` to determine whether the cloud instance exists.